### ISSUE TITLE
Promtail - docker_sd re-sending logs on container shutdown

### ISSUE THREAD

**Issue by heapdavid** [grafana/loki#7103]
**Describe the bug**
Hi,

We're using `docker_sd_config` to scrape docker containers using json-file logging. The symptoms we're seeing when a container shuts down are:
- Recent logs are duplicated once the container has been shut down
- Any older logs cause the following error example:
  - `final error sending batch server returned HTTP status 400 Bad Request (400): entry with timestamp 2022-09-07 08:32:22.028843795 +0000 UTC ignored, reason: 'entry too far behind' for stream: `
  - related to `max_chunk_age`/2 as far as I can tell, as per https://github.com/grafana/loki/issues/5936#issuecomment-1102921754

**To Reproduce**
1. Configure Promtail (v2.6.1) with docker_sd_config as per config below
2. Run a docker container (for longer than `max_chunk_age`/2 since the most recent log entry if you want to reproduce the 2nd symptom)
3. Check the logs don't contain duplicates, then stop or restart the container.
4. Check the logs for the container, which now show duplicates.
5. If long running container as above, check the logs for promtail itself, which shows the 400 error.

**Expected behavior**
Promtail should only send the logs once.

**Environment:**
- Loki & Promtail v2.6.1
- Docker v20.10.17 (reproduced on both Swarm clusters and stand-alone instances)

**Screenshots, Promtail config, or terminal output**
```
  - job_name: docker
    docker_sd_configs:
      - host: unix:///var/run/docker.sock
        refresh_interval: 5s
    relabel_configs:
      - regex: '__meta_docker_container_label_com_(docker_compose_(project|service))'
        replacement: '$1'
        action: labelmap
      - source_labels: [__meta_docker_container_label_com_docker_swarm_service_name]
        target_label: swarm_service
      - source_labels: [__meta_docker_container_label_com_docker_stack_namespace]
        target_label: swarm_stack
      - source_labels: [__meta_docker_container_name]
        regex: '/(.*)'
        target_label: container
    pipeline_stages:
      - docker: {}
      - static_labels:
          job: docker
```

**Comment by heapdavid** [comment]
Just spotted that we get 2 `"finished transferring logs"` messages logged when this happens:

```
Sep 08 12:22:47 [hostname] promtail[597]: level=info ts=2022-09-08T11:22:47.927833034Z caller=target.go:122 target=docker/240580acf5389ea76553252ecefb68cf124d1e2951de5e91481d60d1ca48d5ab msg="finished transferring logs" written=8009 container=240580acf5389ea76553252ecefb68cf124d1e2951de5e91481d60d1ca48d5ab
Sep 08 12:22:48 [hostname] promtail[597]: level=info ts=2022-09-08T11:22:48.160329968Z caller=target.go:122 target=docker/240580acf5389ea76553252ecefb68cf124d1e2951de5e91481d60d1ca48d5ab msg="finished transferring logs" written=8009 container=240580acf5389ea76553252ecefb68cf124d1e2951de5e91481d60d1ca48d5ab
Sep 08 12:22:48 [hostname] promtail[597]: level=error ts=2022-09-08T11:22:48.579512771Z caller=client.go:380 component=client host=[redacted] msg="final error sending batch" status=400 error="server returned HTTP status 400 Bad Request (400): entry with timestamp 2022-09-07 08:32:22.028843795 +0000 UTC ignored, reason: 'entry too far behind' for stream: [redacted]
```

There's definitely only one promtail instance running:

```
# ps aux | grep [p]romtail
root         597  2.7  1.5 2456620 125544 ?      Ssl  Aug24 603:28 /opt/promtail/promtail -config.file /etc/promtail.yml
```

**Comment by cedsana** [comment]
I'm not sure if this is related but I also have duplicate log entries for my docker container. It's a container that runs every minute so maybe it's the same issue? I see 10 times each log line when I query with Loki.

**Comment by crazyproger** [comment]
getting the same error on workload container restart. After restart promtail start to send previously sended logs.

image: grafana/promtail:2.8.0

Is there any workaround maybe?

**Comment by hterik** [comment]
There is a **docker driver** one can use instead of `docker_sd_configs`. [0]
Does anyone know which one is preferable to use? Maybe it does not have the issue described here?

https://grafana.com/docs/loki/next/send-data/docker-driver/

**Comment by jeschkies** [comment]
> Does anyone know which one is preferable to use?

The docker driver has the problem that it locks during a shutdown if it cannot reach Loki. I'd use docker_sd or scrape the log files if Promtail has the permissions.

Regarding the bug I'd need to take a closer look.

**Comment by cstyan** [comment]
this should be fixed via #11563
